allensdk/brain_observatory/ecephys/current_source_density/_current_source_density.py

```diff
@@ -288,3 +283,52 @@ def resample_to_regular(lfp, channels=None, missing_channels=None):
     logging.info('interpolated lfp to {} channels and {} samples'.format(*new_lfp.shape))
     logging.info('total channels: {}'.format(repr(total_channels)))
     return new_lfp, np.array(total_channels)
+
+    
+def compute_csd(accumulated_lfp: np.ndarray, channels: Optional[np.ndarray] = None,
+                missing_channels: Optional[MissingChannelMapping] = None,
+                spacing=0.04) -> Tuple[np.ndarray, np.ndarray]:
+    '''Compute current source density for a subset of channels on a neuropixels probe.
+    
+    Parameters
+    ----------
+    accumulated_lfp : numpy.ndarray
+        Stack of extracted LFP traces surrounding presentation of a common stimulus. Dimensions are trials X channels X samples.
+    channels : Optional[numpy.ndarray], optional
+        Channels to compute CSD for, by default None
+    missing_channels : Optional[MissingChannelMapping], optional
+        Missing channels that need to be addressed by `resample_to_regular()`,
+        by default None
+    spacing : float, optional
+        Vertical spacing between channels, in millimeters, by default 0.04.
+        TODO: spacing needs to be inferred through LFP channel id spacing
+    
+    Returns
+    -------
+    Tuple[csd, channels]:
+        csd : numpy.ndarray
+            Current source density. Dimensions are channels X samples. May
+            contain CSDs for interpolated channels if
+            `missing_channels` was provided.
+        channels: numpy.ndarray
+            Array of channel indices. May contain interpolated channels if
+            `missing_channels` was provided.
+    '''
+
+    accumulated_lfp = accumulated_lfp - np.mean(accumulated_lfp, axis=2)[:, :, None]
+    lfp = np.mean(accumulated_lfp, axis=0)
+
+    lfp, channels = resample_to_regular(lfp, channels, missing_channels)
+
+    lfp_smooth=np.zeros([np.shape(lfp)[0] - 2, np.shape(lfp)[1]])
+    csd = np.zeros([np.shape(lfp_smooth)[0] - 2, np.shape(lfp_smooth)[1]])
+    
+    for t in range(np.shape(lfp)[1]):  # time
+        for d in range(1, np.shape(lfp)[0] - 1):  # depth
+            lfp_smooth[d - 1, t] = 0.25 * ( lfp[d - 1, t] + 2 * lfp[d, t] + lfp[d + 1, t] )
```

Comment:
Smoothing can be similarly vectorized. Perhaps, it would have been better to use one of scipy's smoothing function. One may ask, why use this function. Smoothing can drastically alter CSD results, thus it would help to allow easily modifying smoothing function.